### PR TITLE
[next] improve zod schemas

### PR DESCRIPTION
**Pull request by ItsMeBrianD**
### Description

- Should validate only the first row of a returned dataset (reduce memory usage, improve performance)
- More stringent row checking (columnTypes must match the returned object keys now)

### Checklist

- [x] I have added a [changeset](https://github.com/evidence-dev/evidence/blob/main/CONTRIBUTING.md#adding-a-changeset)

### CODE REVIEW

File: packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js [hunk 1]
```diff
@@ -1,15 +1,90 @@
...
+export const QueryResultSchema = z
+	.object({
+		// Note that this only validates that the first item in the array
+		// is a record with string keys. If the connector returns some
+		// inconsistent array (e.g. [{}, 1]), it will not detect the
+		// invalid row.
+		rows: z.array(z.any()).refine(
+			(data) => {
+				// result is not an array, fail
+				if (!Array.isArray(data)) return false;
```
Comment (Winterhart) [inline packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js:12]: Another way would be to have the `z.any()...` as another `zod` object..

e.g.

```js

rows: z.array(rowItem)

//...

export const rowItemModal = z.record(z.string), z.any())
export type rowItem = z.infer<typeof rowItemModal>;
```

**Review by csjh — APPROVED** [review]
LGTM, though I think we should get rid of `columnTypes` very soon; I don't think inferred Dates would survive the parquet transformation, and after nothing's inferred we can just use `typeof` to identify types

File: packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js [hunk 2]
```diff
@@ -1,15 +1,90 @@
...
+				// result has no rows, we can't validate this
+				// but this is a correct result set
+				if (data.length === 0) return true;
+				try {
+					z.record(z.string(), z.any()).parse(data[0]);
```
Comment (Winterhart) [inline packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js:17]: There's also a `safeParse` equivalent of `parse` but doesn't throw `Error`

File: packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js [hunk 3]
```diff
@@ -1,15 +1,90 @@
...
+				}
+				return true;
+			},
+			{ message: 'Data connector returned invalid rows' }
+		),
+		columnTypes: z.array(
+			z.object({
+				name: z.string(),
+				evidenceType: z.enum(['boolean', 'number', 'string', 'date']),
+				typeFidelity: z.union([z.literal('precise'), z.literal('inferred')])
```
Comment (Winterhart) [inline packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js:29]: There's also the method `.nativeEnum`

File: packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js [hunk 4]
```diff
@@ -1,15 +1,90 @@
...
+	})
+	.refine(
+		(data) => {
+			// Validate that all columnTypes appear
+			if (data.rows.length) {
+				// Filter to column types where name is not in row
+				// Then map columnTypes to their names to make things easier
+				// If there are any columns that were not filtered out; provide an error to zod
+				const missingColumns = data.columnTypes
+					.filter((ct) => !(ct.name in data.rows[0]))
```
Comment (Winterhart) [inline packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js:41]: What's a `ct`?

**Review by Winterhart — COMMENTED** [review]
Looks great! only few questions

File: packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js (hunk 1 above)
Comment (ItsMeBrianD) [reply packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js:12]: I tried that initially and that didn't prevent it from validating every row, doing `z.any()` and then checking the first row in the refine method was the only thing that prevented the full scan

File: packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js (hunk 3 above)
Comment (ItsMeBrianD) [reply packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js:29]: We can make that change if/when we clean up the types in `db-commons`, for now I don't think that value isn't really used so we just want to make sure that it matches the expected interface until we change it

File: packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js (hunk 4 above)
Comment (ItsMeBrianD) [reply packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js:41]: `ct` is a columnType

File: packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js [hunk 5]
```diff
@@ -1,15 +1,90 @@
...
+				if (!Array.isArray(data)) return false;
+				// result has no rows, we can't validate this
+				// but this is a correct result set
+				if (data.length === 0) return true;
+				try {
+					z.record(z.string(), z.any()).parse(data[0]);
+				} catch {
+					return false;
+				}
+				return true;
```
Comment (ItsMeBrianD) [inline packages/plugin-connector/src/data-sources/schemas/query-runner.schema.js:21]:
```suggestion
				return z.record(z.string(), z.any()).safeParse(data[0]).success;
```

**Pull request merged**